### ISSUE TITLE
Remove usage of boost::regex from Scylla

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#7843]
This is a spinoff of comments in https://github.com/scylladb/scylla/issues/5273 and https://github.com/scylladb/scylla/pull/4528:

As noted in https://github.com/scylladb/scylla/issues/5273, using `boost::regex` requires libicu, which has a hefty size cost on our resulting package, not to mention two additional dependencies.

As far as I can tell, `boost::regex` is used in two places, and the goal of this issue is to remove it from both:

1. It is used in `test/tools/cql_repl.cc` because of a gcc bug in `std::regex` explained in commit 15ab6a3febbdf01abcc45d1ab9b1f0cb12021607. This bug might not exist in the clang implementation we now use. Also, we can decide to avoid regex altogether - it is fairly trivial to write the comment checker directly as a state machine.

2. It is used in `utils/like_matcher.cc` for reasons explained in https://github.com/scylladb/scylla/pull/4528#discussion_r291989979. like_matcher converts a pattern which includes the SQL-inspired "_" or "%" characters into a regular expression. `std::regex` works on bytes, not UTF-8, which was perceived to be a problem, but isn't in fact a big problem: The `%` pattern can be converted to a `.*` regular expression which is correct even when it works on bytes. The `_` pattern is a bit more difficult - it needs to match a single unicode character which can be one or more bytes. But it is still trivial to implement `%` as a byte-oriented regular expression: It is either a single ASCII character, or a start byte (begins in bits 11) followed by zero or more continuation bytes (begins with bits 10). That's it! We don't need support in the regular expression engine for character classes, UTF-8 validation, or any other fancy UTF-8 knowledge.

CC @kostja @dekimir

**Comment by dekimir** [comment]
> The `%` pattern can be converted to a `.*` regular expression which is correct even when it works on bytes.

I don't think this is true: `.*` can match bytes corresponding to a non-whole number of UTF-8 characters, producing false positives.

> The `_` pattern is a bit more difficult - it needs to match a single unicode character which can be one or more bytes. But it is still trivial to implement `%` as a byte-oriented regular expression: It is either a single ASCII character, or a start byte (begins in bits 11) followed by zero or more continuation bytes (begins with bits 10)

Yet when I search the net for such a regex, the results are always 40+ characters long.  I would _not_ want to be the one debugging that.

**Comment by nyh** [comment]
In https://github.com/scylladb/scylla/pull/4528#discussion_r549500361 @dekimir raised a very good question: "Aren't there multiple encodings for some characters (ie, different byte sequences that represent the same character and must therefore match each other)?".

Well, luckily (or, rather, intentionally), the most obvious cases of multiple encoding for the same character is *not* a problem: Each character can be encoded in a "minimal" encoding and several "overlong" encodings, but the latter is considered invalid UTF-8 (our UTF-8 validators reject it as input).

There is a much more complicated issue (see https://en.wikipedia.org/wiki/Unicode_equivalence) where the same displayed character can be produced by different sets of unicode characters. For example, ñ can be a single unicode character, or two (ASCII n followed by a "combining tilde" character). I don't think there is any expectation that our LIKE implementation supports this, because I don't think our equality check supports this (to support this, you need to do unicode **normalization** on the strings, and we don't).

**Comment by nyh** [comment]
> > The `%` pattern can be converted to a `.*` regular expression which is correct even when it works on bytes.
> 
> I don't think this is true: `.*` can match bytes corresponding to a non-whole number of UTF-8 characters, producing false positives.

It can't - that's the whole beauty of UTF-8 compared to what preceded it! The point is that UTF-8 is self-synchronizing (https://en.wikipedia.org/wiki/Self-synchronizing_code) - continuation bytes are always different (start with bits 10) - so a valid UTF-8 string can never match a middle of a UTF-8 character.

> > The `_` pattern is a bit more difficult - it needs to match a single unicode character which can be one or more bytes. But it is still trivial to implement `%` as a byte-oriented regular expression: It is either a single ASCII character, or a start byte (begins in bits 11) followed by zero or more continuation bytes (begins with bits 10)
> 
> Yet when I search the net for such a regex, the results are always 40+ characters long. I would _not_ want to be the one debugging that.

If you really want to **validate** UTF-8 and look for errors, it is indeed harder. But that's not what we need to do here. We assume the relevant strings have already been validated. We can treat all bytes starting with bits 10 as continuation bytes - we don't need to check if something is the third or fourth continuation byte.

**Comment by dekimir** [comment]
FWIW, I originally implemented like_matcher without regex:
https://github.com/dekimir/scylla/commit/90c2688973b6ec702cee9dce5fd714f120c79d32

The maintainers didn't like it, so I switched to regex.

**Comment by dekimir** [comment]
> a valid UTF-8 string can never match a middle of a UTF-8 character

OK, but the multi-character graphemes are again a problem.  I imagine `n.*` will match `ñ` encoded as `n~`.  And that cannot be explained away by normalization and comparing to the analogous `=` operator.

**Comment by kostja** [comment]
I think we may need to use libicu again to implement case-insensitive indexes. We don't want to rely on glibc for character conversions, to avoid data corruption when collations suddenly change like happened with PostgreSQL https://www.cybertec-postgresql.com/en/icu-collations-against-glibc-2-28-data-corruption/

**Comment by nyh** [comment]
@kostja that's a fair point. If we expect to need more features from ICU - like case folding, or unicode equivalence - then indeed there is no point in dropping it today just to add it again soon. I don't know, however, if we do have plans for such features.

**Comment by nyh** [comment]
> OK, but the multi-character graphemes are again a problem. I imagine `n.*` will match `ñ` encoded as `n~`. And that cannot be explained away by normalization and comparing to the analogous `=` operator.

Why can't it be explained away? As I noted above, a search for `a = "ñ"` will sometimes match and sometimes won't, depending if this was encoded in one or two unicode characters, so why is it a problem that for the same data, `a LIKE "n%"` will sometimes match and sometimes won't?
Note to self: I should write a test for how equality tests work in Cassandra. Maybe they do some sort of normalization on the stored string and the query, and if they do this, maybe we should to?

**Comment by avikivity** [comment]
> 1. It is used in `test/tools/cql_repl.cc` because of a gcc bug in `std::regex` explained in commit [15ab6a3](https://github.com/scylladb/scylla/commit/15ab6a3febbdf01abcc45d1ab9b1f0cb12021607). This bug might not exist in the clang implementation we now use.

The bug is in libstdc++, not gcc, so the switch to clang does not affect it.

**Comment by nyh** [comment]
> > a valid UTF-8 string can never match a middle of a UTF-8 character
> 
> OK, but the multi-character graphemes are again a problem. I imagine `n.*` will match `ñ` encoded as `n~`. And that cannot be explained away by normalization and comparing to the analogous `=` operator.

@dekimir I wrote a test (`cql-pytest/test_utf8.py`, I'll send it in a minute) that demonstrates the current (with ICU) behaivor of our code. It shows what I meant that ordinary lookup (without LIKE) doesn't obey the supposed equivalence between the two representations of eñe (one as one character, the other as two), but it also demonstrates that today's LIKE - full with ICU - also doesn't support it: "n%" only matches the two-character representation, and "_" only matches the one-character representation. So using ICU didn't get us this feature - if it was ever supposed to.

**Comment by dekimir** [comment]
Good thinking @nyh, thanks.  If Cassandra matches `n%` to `ñ`, that's enough of an explanation.  I now agree it's worth trying `std::regex`, though Kostja is right that we'll need ICU for case-insensitivity (tracked in #4911, btw).

**Comment by nyh** [comment]
Ironically, in https://github.com/scylladb/scylladb/pull/13452 by @denesb we went in the complete opposite direction then what I asked to do here - we switched all the uses of std::regex to boost::regex. 
Supposedly, this was done because std::regex has lousy performance and boost::regex is better. I don't know if this problem still exists two years later, but in any case, the proposal in this issue is as good as dead, so I'll close this issue.

**Comment by denesb** [comment]
> Supposedly, this was done because std::regex has lousy performance and boost::regex is better. I don't know if this problem still exists two years later, but in any case, the proposal in this issue is as good as dead, so I'll close this issue.

The problem is not performance -- the problem is stack usage. Apparently, std::regex uses recursive programming to implement matches and on multiple occasions, this has resulted in stack overflow crashes in production and testing.

**Comment by nyh** [comment]
> The problem is not performance -- the problem is stack usage. Apparently, std::regex uses recursive programming to implement matches and on multiple occasions, this has resulted in stack overflow crashes in production and testing.

Thanks. I knew this problem and even reported it five years ago in https://gcc.gnu.org/bugzilla/show_bug.cgi?id=93502  and lamented it again two years ago in https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86164#c15 - but apparently I blocked it out :-)
